packages/expo-constants/src/Constants.ts

```diff
@@ -27,23 +29,52 @@ if (!ExponentConstants) {
   );
 }
 
-// On Android we pass the manifest in JSON form so this step is necessary
 let manifest = null;
-if (ExponentConstants && ExponentConstants.manifest) {
+// If expo-updates defines a non-empty manifest, prefer that one
+if (NativeModulesProxy.ExpoUpdates) {
+  let updatesManifest;
+  if (NativeModulesProxy.ExpoUpdates.manifest) {
+    updatesManifest = NativeModulesProxy.ExpoUpdates.manifest;
+  } else if (NativeModulesProxy.ExpoUpdates.manifestString) {
+    updatesManifest = JSON.parse(NativeModulesProxy.ExpoUpdates.manifestString);
+  }
+  if (updatesManifest && Object.keys(updatesManifest).length > 0) {
+    manifest = updatesManifest;
+  }
+}
+
+// Fall back to ExponentConstants.manifest if we don't have one from Updates
+if (!manifest && ExponentConstants && ExponentConstants.manifest) {
   manifest = ExponentConstants.manifest;
+  // On Android we pass the manifest in JSON form so this step is necessary
   if (typeof manifest === 'string') {
     manifest = JSON.parse(manifest);
   }
 }
 
-const { name, appOwnership, ...constants } = (ExponentConstants || {}) as any;
+const { name, appOwnership, ...nativeConstants } = (ExponentConstants || {}) as any;
 
-export default {
-  ...constants,
+const constants = {
+  ...nativeConstants,
   // Ensure this is null in bare workflow
   appOwnership: appOwnership ?? null,
-  manifest,
   // Legacy aliases
-  deviceId: constants.installationId,
-  linkingUrl: constants.linkingUri,
-} as Constants;
+  deviceId: nativeConstants.installationId,
+  linkingUrl: nativeConstants.linkingUri,
+};
+
+Object.defineProperties(constants, {
+  manifest: {
+    enumerable: true,
+    get() {
+      if (!manifest) {
+        console.warn(
+          "Constants.manifest is null or undefined. If you're using the bare workflow, this means the embedded config could not be read. Make sure you have installed the expo-constants build scripts."
```

Comment:
can we use `executionEnvironment` here to provide a more context specific message depending on the environment?

for example, if execution environment is bare workflow then:

> "Constants.manifest is null or undefined because the embedded app.config could not be read. Make sure you have installed the expo-constants build scripts."

and if this is the managed workflow, this would likely be a case where we throw an error instead of warning